### PR TITLE
session: preliminary implementation 2

### PR DESCRIPTION
**Pull request by hzhou**
## Pull Request Description
This is 2nd split from https://github.com/pmodels/mpich/pull/4949/commits to simplify review. 

The commits were refactored a bit to keep the tracking of world init state localized as static variables.

[skip warnings]

## Author Checklist
* [x] Reference appropriate issues (with "Fixes" or "See" as appropriate)
* [x] Remove xfail from the test suite when fixing a test
* [x] Commits are self-contained and do not do two things at once
* [x] Commit message is of the form: `module: short description` and follows [good practice](https://chris.beams.io/posts/git-commit/)
* [x] Passes whitespace checkers
* [x] Passes warning tests
* [ ] Passes all tests
* [x] Add comments such that someone without knowledge of the code could understand
* [x] You or your company has a signed contributor's agreement on file with Argonne
* [x] For non-Argonne authors, request an explicit comment from your companies PR approval manager

### CODE REVIEW

**Comment by hzhou** [comment]
test:mpich/ch3/most
test:mpich/ch4/most

**Comment by hzhou** [comment]
@minsii This is a second split from #4949. Could your review it?

**Comment by hzhou** [comment]
test:mpich/ch3/most
test:mpich/ch4/most

**Comment by hzhou** [comment]
test:mpich/ch4/most
test:mpich/ch3/most

**Comment by hzhou** [comment]
@minsii This PR is ready for your review. I have addressed your concern of confusing world state by refactoring to bring the definition and usage of world states localized and wrapped. Both init_counter and init_lock are have become local static variables. The changes to the original mpich_state now are minimized.

File: src/mpi/errhan/errutil.c [hunk 1]
```diff
@@ -197,16 +197,10 @@ void MPII_Errhandler_set_fc(MPI_Errhandler errhand)
 /* Special error handler to call if we are not yet initialized, or if we
    have finalized */
 /* --BEGIN ERROR HANDLING-- */
-void MPIR_Err_preOrPostInit(void)
+void MPIR_Err_Uninitialized(void)
```
Comment (hzhou) [inline src/mpi/errhan/errutil.c]: This function is now renamed

**Comment by minsii** [comment]
The PR looks good to me. I have only a few comments regarding commit separation/message.

File: src/mpi/init/Makefile.mk [hunk 2]
```diff
@@ -5,8 +5,8 @@
 
 mpi_core_sources += \
     src/mpi/init/init_impl.c      \
+    src/mpi/init/init_util.c      \
     src/mpi/init/mpir_init.c      \
-    src/mpi/init/mpir_finalize.c  \
     src/mpi/init/globals.c        \
```
[the remark this answers is not in the document]
Comment (hzhou) [reply src/mpi/init/Makefile.mk:13]: Thanks, done.

File: src/mpi/init/mpi_init.h [hunk 3]
```diff
@@ -61,6 +61,49 @@ int MPII_finalize_async(void);
...
+/* Possible values for process world_model_state */
+#define MPICH_WORLD_MODEL_UNINITIALIZED 0
+#define MPICH_WORLD_MODEL_INITIALIZED   1
+#define MPICH_WORLD_MODEL_FINALIZED     2
+
+extern MPL_atomic_int_t MPIR_world_model_state;
+
+static inline void MPII_world_set_initilized(void)
+{
+    MPL_atomic_store_int(&MPIR_world_model_state, MPICH_WORLD_MODEL_INITIALIZED);
```
[the remark this answers is not in the document]
Comment (hzhou) [reply src/mpi/init/mpi_init.h:79]: Thanks, done.

File: src/include/mpichconfconst.h [hunk 4]
```diff
@@ -60,10 +60,8 @@
...
 
 /* Possible values for process state */
-#define MPICH_MPI_STATE__PRE_INIT 0
-#define MPICH_MPI_STATE__POST_INIT 1
-#define MPICH_MPI_STATE__POST_FINALIZED 2
-/* Internal init states. */
-#define MPICH_MPI_STATE__IN_INIT 3      /* can call MPIR_Err_return_comm(...) on error */
+#define MPICH_MPI_STATE__UNINITIALIZED     0
+#define MPICH_MPI_STATE__MPIR_INITIALIZED  1
+#define MPICH_MPI_STATE__INITIALIZED       2
```
[the remark this answers is not in the document]
Comment (hzhou) [reply src/include/mpichconfconst.h:66]: Your understanding is correct. Commit message updated with your comments.

**Comment by hzhou** [comment]
test:mpich/ch4/most
test:mpich/ch3/most

**Review by minsii — APPROVED** [review]

**Pull request merged**